File: primos.erl
%%% File    : primos.erl
%%% Author  : Jose Castro <jose.r.castro@gmail.com>
%%% Description : 
%%% Created :  9 Sep 2009 by Jose Castro <>

-module(primos).

-export([gen_keys/1, map/1, reduce/1]).

% gen_keys ---------------------------------------------------------
gen_keys({NumStrips,Maximum}) -> gen_strips(2, Maximum, 0, NumStrips, []).

gen_strips(_, _, NumStrips, NumStrips, Result) -> Result;
gen_strips(I, Max, Strip, NumStrips, Result) ->
    Rem  = (Max-2) rem NumStrips,
    Size = if
	       Strip =< Rem -> (Max-2) div NumStrips + 1;
	       true         -> (Max-2) div NumStrips
	   end,
    gen_strips(I+Size, Max, Strip+1, NumStrips, [{I,I+Size-1}|Result]).

% map ---------------------------------------------------------------
map({Min, Max}) ->
    gen_primes(Min, Max, []).

gen_primes(I, Max, Primes) when I > Max -> Primes;
gen_primes(I, Max, Primes) ->
    case is_prime(I) of
	true  -> gen_primes(I+1, Max, [I|Primes]);
	false -> gen_primes(I+1, Max, Primes)
    end.

is_prime(I) -> is_prime(2, math:sqrt(I), I).

is_prime(I, Sqrt, _) when I > Sqrt -> true;
is_prime(I, Sqrt, Num) ->
    if
	(Num rem I) =:= 0 ->
	    false;
	true ->
	    is_prime(I+1, Sqrt, Num)
    end.

% reduce ------------------------------------------------------------
reduce(List) -> lists:sort(lists:append(lists:map(fun({_,X}) -> X end, List))).

File: map_reduce/lib/primos.ex
defmodule Primes do
  @doc """
  ## Examples
    iex(12)> Primes.gen_keys({4,100})
    [{59, 82}, {40, 58}, {21, 39}, {2, 20}]
  """
  def gen_keys({num_strips, maximum}), do: gen_strips(2, maximum, 0, num_strips, [])

  defp gen_strips(_, _, num_strips, num_strips, result), do: result

  defp gen_strips(i, max, strip, num_strips, result) do
    reminder = rem(max - 2, num_strips)

    sixe =
      case strip <= reminder do
        true -> div(max - 2, num_strips + 1)
        false -> div(max - 2, num_strips)
      end

    gen_strips(i + sixe, max, strip + 1, num_strips, [{i, i + sixe - 1} | result])
  end

  @doc """
  ## Examples
    iex(12)> Primes.map({0,10})
    [7, 5, 3, 2, 1, 0]
  """
  def map({min, max}) do
    gen_primes(min, max, [])
  end

  defp gen_primes(i, max, primes) when i > max, do: primes

  defp gen_primes(i, max, primes) do
    case is_prime(i) do
      true -> gen_primes(i + 1, max, [i | primes])
      false -> gen_primes(i + 1, max, primes)
    end
  end

  defp is_prime(i), do: is_prime(2, :math.sqrt(i), i)

  defp is_prime(i, sqrt, _) when i > sqrt, do: true

  defp is_prime(i, sqrt, num) do
    if rem(num, i) == 0 do
      false
    else
      is_prime(i + 1, sqrt, num)
    end
  end

  @doc """
  ## Examples
    iex(10)> Primes.reduce([{59, 82}, {40, 58}, {21, 39}, {2, 20}])
    [2, 21, 40, 59]
  """

  def reduce(list) do
    Enum.reduce(list, [], fn x, acc -> [elem(x, 0) | acc] end)
  end
end
